NonBayesian/Second/second2.py

```diff
@@ -0,0 +1,226 @@
+import numpy as np
+import matplotlib.pyplot as plt
+from timeit import default_timer as timer
+
+
+def srnd_with_zeros(arr):
+    """
+    >>> srnd_with_zeros(np.ones(10))
+    array([0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.])
+    >>> srnd_with_zeros(np.zeros(10))
+    array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+    >>> srnd_with_zeros(np.ones((3, 3)))
+    array([[0., 0., 0., 0., 0.],
+           [0., 1., 1., 1., 0.],
+           [0., 1., 1., 1., 0.],
+           [0., 1., 1., 1., 0.],
+           [0., 0., 0., 0., 0.]])
+    """
+    if arr.ndim == 1:
+        n = np.shape(arr)[0]
+        z_arr = np.zeros(n + 2)
+        z_arr[1:n+1] = arr
+        return z_arr
+
+    if arr.ndim == 2:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        z_arr = np.zeros((n + 2, m+2))
+        z_arr[1:n + 1, 1:m+1] = arr
+        return z_arr
+
+    if arr.ndim == 3:
+        n = np.shape(arr)[0]
+        m = np.shape(arr)[1]
+        k = np.shape(arr)[2]
+        z_arr = np.zeros((n + 2, m + 2, k + 2))
+        z_arr[1:n + 1, 1:m + 1, 1:k + 1] = arr
+        return z_arr
+
+    raise Exception("Bad arguments")
+
+
+def cache(arr):
+    """
+    >>> cache(np.ones(10))
+    array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])
+    >>> cache(np.ones((3, 3)))
+    array([[1., 2., 3.],
+           [2., 4., 6.],
+           [3., 6., 9.]])
+    """
+    cache_arr = np.zeros(np.shape(arr))
+
+    if arr.ndim == 1:
+        for i in range(0, np.shape(arr)[0]):
+            cache_arr[i] = np.sum(arr[0:i+1])
+        return cache_arr
+
+    if arr.ndim == 2:
+        for i in range(0, np.shape(arr)[0]):
+            for j in range(0, np.shape(arr)[1]):
+                cache_arr[i, j] = np.sum(arr[0:i+1, 0:j+1])
+        return cache_arr
+
+    if arr.ndim == 3:
+        for i in range(0, np.shape(arr)[0]):
+            for j in range(0, np.shape(arr)[1]):
+                for k in range(0, np.shape(arr)[2]):
+                    cache_arr[i, j, k] = np.sum(arr[0:i+1, 0:j+1, 0:k+1])
+        return cache_arr
+
+    else:
+        raise Exception("Bad dimension")
```

Comment:
by the way, this thing should be moved left if you remove the `else` above
```suggestion
    raise Exception("Bad dimension")
```